### DISCUSSION TITLE
How to connect to a node that is in other network apart from my local network

### DISCUSSION THREAD

**Discussion by hari-mohan-choudhary** [libp2p/rust-libp2p#2749] [Q&A]
Hi I am able to connect the peers in the same network but I am not able to connect in different  network. How I can do ?

**Comment by mxinden** [comment]
I would need more information in order to be able to help you here. Where is that other node located? What software is it running? What error messages are you seeing? ...

**Comment by hari-mohan-choudhary** [comment]
I run the chat example. I run one node using my wifi and other node I run using my mobile network. I have checked the ip's on the google"what's my ip" and I used these ip. if I connect in the same network It's working fine but when I connect two different service provider network It's not connecting. I am not getting any error. I am running both nodes on two different  macos. Is it any way How I can connect ?

**Reply by thomaseizinger** [reply]
Both, your wifi router and your mobile ISP are operating a NAT.

You need to configure port forwarding, perform hole punching or relay all communication via a relay server.

I am not sure what your knowledge level on these topics is but these key words should get you started on a research journey :)

**Reply by Strange21** [reply]
Hey, Thanks for the suggestion, 
I tried the port forwarding thing, and it is working for me, I am able to receive the ping from my another computer from another network.
But I am wondering why is it that my another machine doesn't need to forward the port in case of receiving the **PONG** messages?
How my (port forwarded) machine is able to send the PONG messages to the machine who didn't configured the **Port forwarding**?

Is it like the router is routing the messages to the ports from where PING originated?

**Reply by haider-rs** [reply]
> Hey, Thanks for the suggestion, I tried the port forwarding thing, and it is working for me, I am able to receive the ping from my another computer from another network. But I am wondering why is it that my another machine doesn't need to forward the port in case of receiving the **PONG** messages? How my (port forwarded) machine is able to send the PONG messages to the machine who didn't configured the **Port forwarding**?
> 
> Is it like the router is routing the messages to the ports from where PING originated?

you can have a read on it here:
https://blog.ipfs.tech/2022-01-20-libp2p-hole-punching/#firewalls-in-a-nutshell

**Reply by thomaseizinger** [reply]
> Hey, Thanks for the suggestion, I tried the port forwarding thing, and it is working for me, I am able to receive the ping from my another computer from another network. But I am wondering why is it that my another machine doesn't need to forward the port in case of receiving the **PONG** messages? How my (port forwarded) machine is able to send the PONG messages to the machine who didn't configured the **Port forwarding**?
> 
> Is it like the router is routing the messages to the ports from where PING originated?

TCP connections (which I assume is what you are using) are bi-directional once set up. In other words, once a machine A successfully connected to an open port on machine B, both ends can send and receive data until the connection is closed.

Similarly, this is why you can browse the web without opening any ports. It is the webserver that needs to be publicly reachable but once the connection is established (by your computer), the webserver can respond to HTTP requests and send the webpage etc back.

**Reply by Strange21** [reply]
Hey, Thanks for explaining. I got the concepts clear now. 
@thomaseizinger and @Haider-Ali-DS
